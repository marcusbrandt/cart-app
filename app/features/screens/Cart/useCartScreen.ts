import { useAppDispatch, useAppSelector } from '../../../store/hooks';
import { removeFromCart } from '../../../store/cartSlice';
import { Product } from '../../types';

const useCartScreen = () => {
  const cart = useAppSelector((state) => state.cart as Product[]);
  const dispatch = useAppDispatch();

  const handleRemoveFromCart = (id: number) => {
    dispatch(removeFromCart(id));
  };

  return {
    cart,
    totalAmount: cart.reduce((acc, item) => acc + item.price, 0),
    handleRemoveFromCart,
  };
};

export default useCartScreen;
