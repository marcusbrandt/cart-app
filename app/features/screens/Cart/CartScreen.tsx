import { Text, View } from 'react-native';
import useCartScreen from './useCartScreen';
import CartProduct from '../../components/CartProduct';

export const CartScreen = () => {
  const { cart, handleRemoveFromCart } = useCartScreen();

  if (!cart || cart.length === 0) {
    return (
      <View style={{ alignSelf: 'center', alignItems: 'center' }}>
        <Text>Carrinho vazio!</Text>
      </View>
    );
  }
  return (
    <View>
      {cart.map((product) => (
        <CartProduct
          key={product.id}
          product={product}
          onClick={handleRemoveFromCart}
        />
      ))}
    </View>
  );
};
