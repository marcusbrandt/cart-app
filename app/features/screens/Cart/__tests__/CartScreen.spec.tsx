import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { CartScreen } from '../CartScreen';
import * as useCartScreen from '../useCartScreen';
import { render } from '../../../../../__tests__/__utils__/wrapperUtils';

const mockUseCartScreen = jest.spyOn(useCartScreen, 'default');
const handleRemoveFromCart = jest.fn();

const mockDefault = {
  cart: [
    {
      id: 1,
      name: 'Product 1',
      price: 10,
      image: 'https://via.placeholder.com/100',
      cart: true,
    },
    {
      id: 2,
      name: 'Product 2',
      price: 20,
      image: 'https://via.placeholder.com/100',
      cart: true,
    },
  ],
  handleRemoveFromCart,
};

const mockEmpty = {
  cart: [],
  handleRemoveFromCart,
};

describe('CartScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseCartScreen.mockReturnValue(mockDefault);
  });

  it('renders empty cart message when cart is empty', () => {
    mockUseCartScreen.mockReturnValue(mockEmpty);
    const { getByText, toJSON } = render(<CartScreen />);
    expect(getByText('Carrinho vazio!')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders products in the cart', () => {
    const { getByText, toJSON } = render(<CartScreen />);
    expect(getByText('Product 1')).toBeTruthy();
    expect(getByText('Product 2')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls handleRemoveFromCart when a product is removed', () => {
    const { getByTestId } = render(<CartScreen />);
    fireEvent.press(getByTestId('cart-product-remove-button-1'));
    expect(handleRemoveFromCart).toHaveBeenCalled();
  });
});
