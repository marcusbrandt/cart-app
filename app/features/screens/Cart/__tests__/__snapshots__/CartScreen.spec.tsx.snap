// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartScreen renders empty cart message when cart is empty 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "alignSelf": "center",
    }
  }
>
  <Text>
    Carrinho vazio!
  </Text>
</View>
`;

exports[`CartScreen renders products in the cart 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 8,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 8,
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://via.placeholder.com/100",
        }
      }
      style={
        {
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "height": 100,
          "objectFit": "cover",
          "width": 100,
        }
      }
      testID="cart-product-image-1"
    />
    <Text
      style={
        {
          "color": "black",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
      testID="cart-product-title-1"
    >
      Product 1
    </Text>
    <Text
      style={
        {
          "color": "black",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
      testID="cart-product-price-1"
    >
      10
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "display": "flex",
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        }
      }
      testID="cart-product-remove-button-1"
    >
      <Image
        source={
          {
            "uri": "https://via.placeholder.com/24",
          }
        }
        style={
          {
            "height": 24,
            "width": 24,
          }
        }
        testID="cart-product-remove-button-image-1"
      />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 8,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 8,
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://via.placeholder.com/100",
        }
      }
      style={
        {
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "height": 100,
          "objectFit": "cover",
          "width": 100,
        }
      }
      testID="cart-product-image-2"
    />
    <Text
      style={
        {
          "color": "black",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
      testID="cart-product-title-2"
    >
      Product 2
    </Text>
    <Text
      style={
        {
          "color": "black",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
      testID="cart-product-price-2"
    >
      20
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "display": "flex",
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        }
      }
      testID="cart-product-remove-button-2"
    >
      <Image
        source={
          {
            "uri": "https://via.placeholder.com/24",
          }
        }
        style={
          {
            "height": 24,
            "width": 24,
          }
        }
        testID="cart-product-remove-button-image-2"
      />
    </View>
  </View>
</View>
`;
